pkg/resource/deploy/step_executor.go

```diff
@@ -0,0 +1,305 @@
+		}
+	}
+
+	events := se.opts.Events
+	if events != nil {
+		if postErr := events.OnResourceStepPost(payload, step, status, err); postErr != nil {
+			se.log(workerID, "step %v on %v failed post-resource step: %v", step.Op(), step.URN(), postErr)
+			se.writeError(step.URN(), "post-step event returned an error: %s", postErr.Error())
+			return false
+		}
+	}
+
+	return true
+}
+
+// log is a simple logging helper for the step executor.
+func (se *stepExecutor) log(workerID int, msg string, args ...interface{}) {
+	if logging.V(stepExecutorLogLevel) {
+		message := fmt.Sprintf(msg, args...)
+		logging.V(stepExecutorLogLevel).Infof("StepExecutor worker(%d): %s", workerID, message)
+	}
+}
+
+// writeError writes an error to the standard diagnostic mechanism, with the given URN and message.
+// The message is anchored to the resource with the given URN's row in the CLI output.
+func (se *stepExecutor) writeError(urn resource.URN, msg string, args ...interface{}) {
+	diagMsg := diag.RawMessage(urn, fmt.Sprintf(msg, args...))
+	se.plan.Diag().Errorf(diagMsg)
+}
+
+//
+// The step executor owns a number of goroutines that it considers to be "workers", responsible for
+// executing steps. By default, as we ease into the waters of parallelism, there is at most one worker
+// active.
+//
+// Workers continuously pull from se.incomingChains, executing chains as they are provided to the executor.
+// There are two reasons why a worker would exit:
+//
+//  1. A worker exits if se.ctx is canceled. There are two ways that se.ctx gets canceled: first, if there is
+//     a step error in another worker, it will cancel the context. Second, if the plan executor experiences an
+//     error when generating steps or doing pre or post-step events, it will cancel the context.
+//  2. A worker exits if it experiences an error when running a step.
+//
+
+// worker is the base function for all step executor worker goroutines. It continuously polls for new chains
+// and executes any that it gets from the channel.
+func (se *stepExecutor) worker(workerID int) {
+	se.log(workerID, "worker coming online")
+	se.workers.Add(1)
+	defer se.workers.Done()
+
+outer:
+	for {
+		se.log(workerID, "worker waiting for incoming chains")
+		select {
+		case chain := <-se.incomingChains:
+			if chain == nil {
+				se.log(workerID, "worker received nil chain, exiting")
+				break outer
+			}
+
+			se.log(workerID, "worker received chain for execution")
+			se.executeChain(workerID, chain)
+		case <-se.cancel.Context().Canceled():
+			se.log(workerID, "worker exiting due to cancellation")
+			break outer
+		}
+	}
+
+	se.log(workerID, "worker terminating")
+}
+
+func newStepExecutor(cancel *cancel.Source, plan *Plan, opts Options, preview bool) *stepExecutor {
+	exec := &stepExecutor{
+		plan:           plan,
+		opts:           opts,
+		preview:        preview,
+		incomingChains: make(chan Chain),
+		cancel:         cancel,
+		sawError:       0,
+	}
+
+	fanout := opts.Parallel
```

Comment:
Slightly subtle that `new`, which is often just a "constructor", spawns concurrent work. It may be worth factoring this bit out into a separate `Start` function. Up to you.

Reply:
`SnapshotManager` also follows this pattern - it spawns a goroutine in its "constructor". Since all users of `stepExecutor` are going to immediately call `Start` after the constructor finishes anyway, I think I'd like to keep the initialization logic all in one function so that it can't be messed up.